### PR TITLE
#267: Create Context for song and playlist currently being played

### PR DESCRIPTION
**Pull request by SaurabhGurde**
## Description

Created a global context for changing song from any component without prop drilling

## Commit type

- `feat`: indicates the addition of a new feature or functionality to the project.

## Issue

#267 

## Solution

created song change context in useSongChangeContextApi.tsx file a custom hook which can used anywhere in any component to change global current song.

## Proposed Changes

- added context provider in app level

## Assigned

@AntonioMrtz

### CODE REVIEW

**Comment by SaurabhGurde** [comment]
@AntonioMrtz created a context not yet used anywhere

File: Electron/src/renderer/App.tsx [hunk 1]
```diff
@@ -63,106 +64,108 @@ function App() {
...
+                refreshSidebarData={refreshSidebarData}
+              />
+              <div
+                className={`App d-flex container-fluid ${styles.mainContentWrapper}`}
+              >
+                <Routes>
+                  <Route
+                    path="/playlist/:id"
+                    element=<Playlist
+                      changeSongName={changeSongName}
```
Comment (AntonioMrtz) [inline Electron/src/renderer/App.tsx:84]: `changeSongName` is still being passed as argument here, we should check for the function usages and use the context instead

File: Electron/src/renderer/App.tsx [hunk 2]
```diff
@@ -63,106 +64,108 @@ function App() {
...
-                />
-                <Route
-                  path="/explorar"
-                  element=<Explorar
-                    changeSongName={changeSongName}
-                    refreshSidebarData={refreshSidebarData}
+                  <Route
+                    path="/explorar"
+                    element=<Explorar
+                      changeSongName={changeSongName}
```
Comment (AntonioMrtz) [inline Electron/src/renderer/App.tsx:91]: Ídem

**Review by AntonioMrtz — CHANGES_REQUESTED** [review]

**Review by AntonioMrtz — CHANGES_REQUESTED** [review]
Replace `changeSongName` passed as parameter (prop drilling) with the use of the context. Also add tests for the new functionality.

Run `npm run lint` because I saw some issues with linting.

**Comment by SaurabhGurde** [comment]
@AntonioMrtz fixed lint issue and remove prop drilling and used custom hook in every component which is using changSongName

File: Electron/src/hooks/useSongChangeContextApi.tsx [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  const changeSongName = (newSongName: string) => {
+    setSongName(newSongName);
+  };
+
+  return (
+    <SongNameContext.Provider value={{ songName, changeSongName }}>
+      {children}
+    </SongNameContext.Provider>
+  );
+}
```
Comment (SaurabhGurde) [inline Electron/src/hooks/useSongChangeContextApi.tsx:32]: @AntonioMrtz value object may create new reference at every re-render of provider so should we use useMemo

File: Electron/src/components/ShowAllItems/types/PropsShowAllItems.ts [hunk 4]
```diff
@@ -9,5 +9,5 @@ export enum ShowAllItemsTypes {
 export interface PropsAllItems {
   refreshSidebarData: () => void;
   type: ShowAllItemsTypes;
-  changeSongName: (songName: string) => void;
+  changeSongName?: (songName: string) => void;
```
Comment (AntonioMrtz) [inline Electron/src/components/ShowAllItems/types/PropsShowAllItems.ts:12]: Why do we keep references to the method? Shouldn't we just use the context? I don't think there's a need for that prop anymore. Check the other props modified

File: Electron/src/hooks/useSongChangeContextApi.tsx (hunk 3 above)
Comment (AntonioMrtz) [reply Electron/src/hooks/useSongChangeContextApi.tsx:32]: Yeah @SaurabhGurde , no problem with that, add the useMemo behaviour

**Review by AntonioMrtz — CHANGES_REQUESTED** [review]
Add useMemo + review failing tests + add tests for the new functionality

**Comment by SaurabhGurde** [comment]
@AntonioMrtz all changes are done as you said. and tested functionality which is working well.

**Comment by AntonioMrtz** [comment]
> @AntonioMrtz all changes are done as you said. and tested functionality which is working well.

Not only test the functionality. But add unit tests to the code about the implemented functionality :)

**Comment by SaurabhGurde** [comment]
@AntonioMrtz All new test are added and unit test are done by my side and they are passing plz check

**Comment by AntonioMrtz** [comment]
Hi @SaurabhGurde , code looks code. Yeah, unit tests are fixed but I'm missing a new unit test that validates whether the playing song's state is being correctly modified. Once the new test is being added to the codebase I will do a last in-depth review of the code and merge if everything is correct

**Comment by SaurabhGurde** [comment]
> Hi @SaurabhGurde , code looks code. Yeah, unit tests are fixed but I'm missing a new unit test that validates whether the playing song's state is being correctly modified. Once the new test is being added to the codebase I will do a last in-depth review of the code and merge if everything is correct

I did not understand plz explain more

**Comment by AntonioMrtz** [comment]
> > Hi @SaurabhGurde , code looks code. Yeah, unit tests are fixed but I'm missing a new unit test that validates whether the playing song's state is being correctly modified. Once the new test is being added to the codebase I will do a last in-depth review of the code and merge if everything is correct
> 
> I did not understand plz explain more

*Code looks good*  hehe, it was a typo, my bad. I was saying that the following is required:

* Write Unit test that validates changing song actually works with the new context. I would use the file `PlaylistUser.test.tsx` and add a test that simulates the user clicking the song. The expected result is the context changing the current song name. Then simulate changing to another song and expect the new songname

After that, I will review the code for a last time and merge the code :)

**Comment by SaurabhGurde** [comment]
@AntonioMrtz added a test but i feel that it may be wrong or missing something plz check

File: Electron/src/components/Cards/SongCard/SongCard.tsx [hunk 5]
```diff
@@ -63,6 +65,7 @@ export default function SongCard({
         className={`rounded ${styles.card}`}
         onDoubleClick={handleClickSong}
         onContextMenu={handleRightClick}
+        data-testid="song-card"
```
Comment (AntonioMrtz) [inline Electron/src/components/Cards/SongCard/SongCard.tsx:68]: All cards will have the same test id :(. We have to find another way of selecting the card

File: Electron/src/__tests__/pages/PlaylistUser.test.tsx [hunk 6]
```diff
@@ -522,3 +519,69 @@ test('Playlist user role update playlist', async () => {
 
   expect(refreshSidebarData).toHaveBeenCalledTimes(1);
 });
+
+test('Playlist updates song name in context when a song is clicked', async () => {
```
Comment (AntonioMrtz) [inline Electron/src/__tests__/pages/PlaylistUser.test.tsx:523]: Test is good, but we need to select the song card in another way (no testid) as said in the other comment

**Review by AntonioMrtz — CHANGES_REQUESTED** [review]

File: Electron/src/__tests__/pages/PlaylistUser.test.tsx [hunk 7]
```diff
@@ -522,3 +519,69 @@ test('Playlist user role update playlist', async () => {
...
+              }
+            />
+          </Routes>
+          <Footer />
+        </SongNameChangeContextProvider>
+      </MemoryRouter>,
+    );
+  });
+
+  const songCard = await screen.findByTestId(`song-card-${songMockFetch.name}`);
```
Comment (SaurabhGurde) [inline Electron/src/__tests__/pages/PlaylistUser.test.tsx:580]: I think this should work added name as unique parameter

File: Electron/src/components/Cards/SongCard/SongCard.tsx [hunk 8]
```diff
@@ -63,6 +65,7 @@ export default function SongCard({
         className={`rounded ${styles.card}`}
         onDoubleClick={handleClickSong}
         onContextMenu={handleRightClick}
+        data-testid={`song-card-${name}`}
```
Comment (SaurabhGurde) [inline Electron/src/components/Cards/SongCard/SongCard.tsx:68]: here same!

**Comment by AntonioMrtz** [comment]
New test is failing, did you check the result before pushing it?

**Comment by SaurabhGurde** [comment]
> New test is failing, did you check the result before pushing it?

Yes I use command npm run lint everytime before pushing. But nom run test fails because we are using dummy data for api testing due to which it gives error "cannot read property map of undefined" because api don't send anything so it tries to apply map on undefined variable

**Comment by AntonioMrtz** [comment]
> > New test is failing, did you check the result before pushing it?
> 
> Yes I use command npm run lint everytime before pushing. But nom run test fails because we are using dummy data for api testing due to which it gives error "cannot read property map of undefined" because api don't send anything so it tries to apply map on undefined variable

Thats the point, mocking the data, if the data is empty ofc is going to fail. Whats the purpose of the test if it fails or only works with the app launched?

**Comment by AntonioMrtz** [comment]
I will do the tests this time, but I'm not accepting any PR without working tests :)

**Review by AntonioMrtz — APPROVED** [review]

**Comment by SaurabhGurde** [comment]
> I will do the tests this time, but I'm not accepting any PR without working tests :)

const artistMockFetch = {
  name: userName,
  photo: 'photo',
  register_date: 'date',
  password: 'hashpassword',
  playback_history: [songName],
  playlists: [playlistName],
  saved_playlists: [playlistName],
  uploaded_songs: [songName],
};

we are using "artistMockFetch.name" this dummy data for to call API to get song list for testing and "songMockFetch.name" to check whether current song being playing is present in footer or not. but its fake data so api dont give any data and test will fail. we have to use real song name data of currently logged in user. Thats what i wanted to ask how to get real data of logged in user to add it in test.tsx file and test api.

**Pull request merged**